### ISSUE TITLE
Consider to remove pre-write actions in C++

### ISSUE THREAD

**Issue by mikir** [ndsev/zserio#342]
The C++ BitStreamWriter does not allocate memory by itself. This allows applications to control memory allocation strategy.

However, this causes that application needs to call `bitSizeOf()` method before `write()` method. So, application must call as well
`initializeChildren()` and `initializeOffsets()` methods. Because of that, pre-write actions in `write()` method which calls automatically also `initializeChildren()` and `initializeOffsets()` are not useful anymore.

Moreover, `write()` method cannot be `const` because of these pre-write actions. Therefore, consider to remove completely pre-write actions.

**Comment by mikir** [comment]
Consider to introduce some new user friendly method which will initialize offsets and children which could be generated in all zserio objects.

**Comment by josephbirkner** [comment]
Maybe we can also think about marking the offset info in C++ as `mutable` to allow for serializing const objects? Or have you already decided against this? 🤔

**Comment by mikir** [comment]
We have decided to remove all pre-write actions completely. This allows to have `write()` method `const` which is expected. Beside of that, `write()` method won't have any side effects anymore which was not expected.

Together with this change, we have improved `serialize()` and `deserialize()` methods in C++ runtime library. Now, the method `serialize()` automatically calls before serialization the following initialization methods:

- `initialize()` method for types with parameters
- `initializeChildren()` method for types which contain types with parameters
- `initializeOffsets()` method for all types
